src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,317 @@
+import org.opensearch.watcher.FileWatcher;
+import org.opensearch.watcher.WatcherHandle;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Represents a persistent allocation made in native memory. In this case, persistent means that the allocation is made
+ * and not freed in the same call to the JNI. Therefore, in order to prevent memory leaks, we need to ensure that each
+ * allocation is properly freed
+ */
+public interface NativeMemoryAllocation {
+
+    /**
+     * Closes the native memory allocation. It should deallocate all native memory associated with this allocation.
+     */
+    void close();
+
+    /**
+     * Check if the allocation has been closed.
+     *
+     * @return true if allocation has been closed; false otherwise
+     */
+    boolean isClosed();
+
+    /**
+     * Get the native memory pointer associated with the native memory allocation.
+     *
+     * @return pointer to native memory allocation
+     */
+    long getPointer();
+
+    /**
+     * Locks allocation for read.
+     */
+    void readLock();
+
+    /**
+     * Locks allocation for write.
+     */
+    void writeLock();
+
+    /**
+     * Unlocks allocation for read.
+     */
+    void readUnlock();
+
+    /**
+     * Unlocks allocation for write.
+     */
+    void writeUnlock();
+
+    /**
+     * Get the size of the native memory allocation in kilobytes.
+     *
+     * @return size of native memory allocation
+     */
+    long getSizeInKb();
+
+    /**
+     * Represents native indices loaded into memory. Because these indices are backed by files, they should be
+     * freed when file is deleted.
+     */
+    class IndexAllocation implements NativeMemoryAllocation {
+
+        private final long pointer;
+        private final long size;
+        private volatile boolean closed;
+        private final KNNEngine knnEngine;
+        private final String indexPath;
+        private final String osIndexName;
+        private final ReadWriteLock readWriteLock;
+        private final WatcherHandle<FileWatcher> watcherHandle;
+
+        /**
+         * Constructor
+         *
+         * @param pointer Pointer in memory to the index
+         * @param size Size this index consumes in kilobytes
+         * @param knnEngine KNNEngine associated with the index allocation
+         * @param indexPath File path to index
+         * @param osIndexName Name of OpenSearch index this index is associated with
+         * @param watcherHandle Handle for watching index file
+         */
+        public IndexAllocation(long pointer, long size, KNNEngine knnEngine, String indexPath, String osIndexName,
+                               WatcherHandle<FileWatcher> watcherHandle) {
+            this.closed = false;
+            this.knnEngine = knnEngine;
+            this.indexPath = indexPath;
+            this.osIndexName = osIndexName;
```

Comment:
nit: indexName

Reply:
I will rename to opensearchIndexName. indexPath refers to the Lucene index. So to disambiguate, I will qualify this index name with opensearch.